host/docs/uhd_find_devices.1
.TH "uhd_find_devices" 1 "3.7.0" UHD "User Commands"
.SH NAME
uhd_find_devices \- USRP Hardware Driver Discovery Utility
.SH DESCRIPTION
Find UHD-supporting Ettus Research products attached by USB,
network or embedded configuration.
.LP
The UHD package is the universal hardware driver for Ettus Research
products. The goal is to provide a host driver and API for
current and future Ettus Research products. Users will be able to use
the UHD driver standalone or with 3rd party applications.
.SH SYNOPSIS
.B  uhd_find_devices [OPTIONS]
.SH OPTIONS
.IP "Device Address Arguments:"
--args \fIarg\fR
.IP "This help information:"
--help
.SH IDENTIFYING DEVICES
.sp
Devices are addressed through key/value string pairs.
These string pairs can be used to narrow down the search for a specific device or group of devices.
Most UHD utility applications and examples have a \-\-args parameter that takes a device address,
where the device address is expressed as a delimited string.
See the documentation in types/device_addr.hpp for reference.

Every device has several ways of identifying it on the host system.
.SS Identifying by hardware identifier
.sp

All Ettus Research devices can be found through their hardware series identifier, which match to
devices as follows:

Argument       |  Device

type=usrp1     |  USRP1

type=usrp2     |  USRP2, USRP N200, USRP N210

type=b100      |  USRP B100

type=b200      |  USRP B200, USRP B210

type=e3x0      |  USRP E310

type=x300      |  USRP X300, USRP X310

type=octoclock |  OctoClock

.SS Identifying by serial number

All Ettus Research devices are given a unique serial number, which can be used to identify a device as follows:

serial=12345678

.SS Identifying by IP address

USRP2, USRP N200, USRP N210, USRP X300, USRP X310, and OctoClock devices connected to the host machine can all be found through their
IP addresses, as follows:

addr=192.168.10.2

.SS Identifying by name

Users have the option of giving their devices short names using the usrp_burn_mb_eeprom and octoclock_burn_eeprom
utilities in lib/uhd/utils. Devices that have been given a name can be identified as follows:

name=foo

.fi
.SH EXAMPLES
.SS Device discovery via command line
.sp
Device address arguments can be supplied to narrow the scope of the search.
.sp
.nf
.ft C
uhd_find_devices \-\-args="type=usrp1"

\-\- OR \-\-

uhd_find_devices \-\-args="serial=12345678"
.ft P
.fi
.IP "Find all devices available to this system:"
uhd_find_devices
.SH SEE ALSO
UHD documentation:
.B http://files.ettus.com/manual/
.LP
GR-UHD documentation:
.B http://gnuradio.org/doc/doxygen/page_uhd.html
.LP
Other UHD programs:
.sp
uhd_usrp_probe(1)
.SH AUTHOR
This manual page was written by Maitland Bottoms and Nicholas Corgan
for the Debian project (but may be used by others).
.SH COPYRIGHT
Copyright (c) 2010 Ettus Research LLC
.LP
This program is free software: you can redistribute it and/or modify
it under the terms of the GNU General Public License as published by
the Free Software Foundation, either version 3 of the License, or
(at your option) any later version.
.LP
This program is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
GNU General Public License for more details.


CHANGELOG
Change Log for Releases
==============================

## 003.012.000.000
* N3x0: Add White Rabbit support, add N300 support, standard BIST
        includes fan, fix issue with 1GigE, switch to 2 radio blocks
        with 2 channels each, upgrade TDC to version 2.0, fix issue in
        ARM deframer
* X300: Enable BasicRX to use A/B/AB/BA muxing setups, more consistent
        logging, fix enumeration issue with TwinRX
* USRP2/N2x0: Re-add ability to modulate in the DAC, improve ISE
              settings to better meet timing
* B205mini: Fix global reset, improve timing in b205_ref_pll
* UHD: Remove a lot of Boost usage, mostly replaced by C++11 features,
       more unit tests, fix Boost 1.67 compatibility, fix compiler
       warnings, add API to query clock rate range, fix get_usrp_?x_info
* MPM: Refactored N3xx code, moved C++ standard to 14, refactor
       Boost.Python bindings, use CMake variable MPM_DEVICE
* Logging: Allow disabling fastpath msgs at runtime
* Docs: Clarified meaning of DSP frequencies, improved manual
        section on synchronization, added some known issues to B100,
        USRP2, and USRP1, update test test procedure description
* Examples: Improved benchmark_rate (added failure thresholds, fixed
  incorrect calculation of samples on drops, fixed timeout values),
  minor fixes to txrx_loopback_to_file
* Utils: Handle U's in calibration tools, create-lvbitx.py is now Py3k
         compatible, fixed git-hash.sh
* RFNoC: DDCs/DUCs use DDSes instead of CORDIC, add DMA-based replay
         block in FPGA, add 64-bit support to axi_wrapper, add compat
         number to radio block,
* Debian: Fix rules file, fix Changelog format
* Fix license headers
* This release includes all bugfixes and features from previous
  releases, in particular, the 3.11.* release cycle
* Known issues: N310: In some rare scenarios at high rates, streaming
  can fail and even bring the FPGA into a bad state.

## 003.011.001.000
* N310: fix compiler warnings
* N310: Implement digital loopback
* N3xx: Add N3xx devtest
* X300: Properly coerce master clock rate (tick rate)
* X300: handle bad weak_ptr during pcie discovery
* X300: handle bad weak_ptr during pcie discovery
* X300: Fix check_radio_config() to fix errors when using a single dboard
        in slot A
* B200: docs: Suggest modifying recv_frame_size for more stability
* B200: Fix bandwidth warnings and ranges
* N2xx: Fix regression issue that limited tuning range
* UBX: Change antenna functions to coercers on antenna/value properties
* adf4002: Fix register programming for power down bit
* UHD: Fix config file path for some Windows builds
* UHD: Add operators == and != for uhd::dict
* UHD: Add device_addr_t constructor from map
* UHD: Fix range of gain group to skip gains with zero step
* UHD: Changes to support Boost 1.67
* UHD: Correctly set end of burst flag in RX metadata
* UHD: Reduce usage of boost::assign, boost::this_thread::sleep, and boost:bind
* UHD: Update multi_usrp::get_usrp_?x_info() for MPM devices
* UHD: Refactor static const values to fix linker errors in niusrprio
* mpm: cmake: Add git hash and version info to Python module
* mpm: Add reference counters to UIO
* mpm: Add offset to EEPROM reads
* mpm: Disable PPS out during initialization
* mpm: Update cmake to find the correct python3
* mpm: Bump maximum supported revision to 5 (Rev F)
* mpm: Fixed db slot typo in db-id
* mpm: Increased claim timeout, made a separate RPC connection for claim, and
       added asyn calls for long RPC executions
* mpm: Improve xport<->SFP mapping algorithm
* mpmd: Improved find routine to fail fast and verify correct device is
        reachable
* mpmd: Add missing virtual destructors
* rfnoc/x300: Make sure peek32() and peek64() are called with actual addresses
* rfnoc: ctrl_iface cleanup
* rfnoc radio: Improve warning for too many samples requested
* rfnoc radio: get_rx_stream resets sequence num
* examples: Increase settling time, increase buffer fill time, and fix subdevice
            selection in txrx_loopback_to_file
* examples: Improvements to benchmark_rate
* utils: downloader supports multiple RegExs
* utils: Added code to handle underruns during self calibration
* utils: Fix 30s tiemout in query_gpsdo_sensors
* logging: Improve style consistency and demote some messages
* logging: Fix UHD_LOG_FILE cmake variable
* Docs: Add Known Issues section to USRP1, B100, and USRP2/N2x0
* Docs: Hide dependencies directory from Doxygen
* Docs: Clarify subdev specs and magnesium driver usage for N300/N310
* cmake: Improve warning for missing requests
* cmake: update NSIS template
* cmake: Remove images downloader section (replaced by manifest)

## 003.011.000.001
* N300: Added driver support (includes some refactoring of N3xx
  codebase)
* MPM: Fix PyLint warnings, Fix error handling for TCA communication
  errors, Fix printout of AD9371 version
* uhd_images_downloader: Create unique archive names for images archives
  (now include git hashes in the filename)
* uhd_images_downloader: Fix SHA256 check
* utils: Add support for N3xx filesystem images to images downloader
* UHD: Minor logging fixes
* UHD: fix legacy compat to work with 2TX radio block
* X300: improve lvbitx bitstream md5 read time
* examples: Enhance benchmark_rate with more stats and timestamps for errors
* cmake: Correctly fail when an unavailable component is requested
* debian: Add UHDConfig.cmake to install list for libuhd-dev

## 003.011.000.000
* N310: Added driver
* UBX: Add support for CAL antenna for Rev E
* Added the module peripheral manager (MPM) with all subcomponents (N310
  drivers, mpm_shell, RPC server, BIST, etc.)
* UHD: Added rpclib as a internally tracked dependency
* UHD: Reduced the usage of Boost
* UHD: Updated uhd_images_downloader: Now uses more elaborate manifest to
  optimize downloads
* UHD: Introduced uhdlib internal include paths
* UHD: Add support for configuration files. USRP settings can now be set using

host/utils/uhd_find_devices.cpp
//
// Copyright 2010 Ettus Research LLC
// Copyright 2018 Ettus Research, a National Instruments Company
//
// SPDX-License-Identifier: GPL-3.0-or-later
//

#include <uhd/utils/safe_main.hpp>
#include <uhd/device.hpp>
#include <boost/program_options.hpp>
#include <boost/format.hpp>
#include <iostream>
#include <cstdlib>

namespace {
    //! Conditionally append find_all=1 if the key isn't there yet
    uhd::device_addr_t append_findall(const uhd::device_addr_t& device_args)
    {
        uhd::device_addr_t new_device_args(device_args);
        if (!new_device_args.has_key("find_all")) {
            new_device_args["find_all"] = "1";
        }

        return new_device_args;
    }
}

namespace po = boost::program_options;

int UHD_SAFE_MAIN(int argc, char *argv[]){
    po::options_description desc("Allowed options");
    desc.add_options()
        ("help", "help message")
        ("args", po::value<std::string>()->default_value(""), "device address args")
    ;

    po::variables_map vm;
    po::store(po::parse_command_line(argc, argv, desc), vm);
    po::notify(vm);

    //print the help message
    if (vm.count("help")){
        std::cout << boost::format("UHD Find Devices %s") % desc << std::endl;
        return EXIT_SUCCESS;
    }

    //discover the usrps and print the results
    const uhd::device_addr_t args(vm["args"].as<std::string>());
    uhd::device_addrs_t device_addrs =
        uhd::device::find(append_findall(args));
    if (device_addrs.empty()) {
        std::cerr << "No UHD Devices Found" << std::endl;
        return EXIT_FAILURE;
    }

    typedef std::map<std::string, std::set<std::string> > device_multi_addrs_t;
    typedef std::map<std::string, device_multi_addrs_t> device_addrs_filtered_t;
    device_addrs_filtered_t found_devices;
    for (auto it = device_addrs.begin(); it != device_addrs.end(); ++it) {
        std::string serial = (*it)["serial"];
        found_devices[serial] = device_multi_addrs_t();
        for(std::string key: it->keys()) {
            if (key != "serial") {
                found_devices[serial][key].insert(it->get(key));
            }
        }
        for (auto sit = it + 1; sit != device_addrs.end();) {
            if ((*sit)["serial"] == serial) {
              for(std::string key: sit->keys()) {
                    if (key != "serial") {
                        found_devices[serial][key].insert(sit->get(key));
                    }
                }
                sit = device_addrs.erase(sit);
            } else {
                sit++;
            }
        }
    }

    int i = 0;
    for (auto dit = found_devices.begin(); dit != found_devices.end(); ++dit) {
        std::cout << "--------------------------------------------------"
                  << std::endl;
        std::cout << "-- UHD Device " << i << std::endl;
        std::cout << "--------------------------------------------------"
                  << std::endl;
        std::stringstream ss;
        ss << "Device Address:" << std::endl;
        ss << boost::format("    serial: %s") % dit->first << std::endl;
        for (auto mit = dit->second.begin(); mit != dit->second.end(); ++mit) {
            for (auto vit = mit->second.begin(); vit != mit->second.end();
                 ++vit) {
                ss << boost::format("    %s: %s") % mit->first % *vit
                   << std::endl;
            }
        }
        std::cout << ss.str() << std::endl << std::endl;
        i++;
    }

    return EXIT_SUCCESS;
}
